ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -372,14 +372,19 @@ public boolean hasFailedOnNotFound() {
     // To account for GET operation, the threshold should be  >= totalReplicaCount - (success target - 1)
     // Right now, this only applies for disk replica only partitions and may not be completely accurate if there are
     // failures responses other than not found.
-    // TODO support cloud replicas in this condition, also account for failures other than not found
-    if (crossColoEnabled && !cloudReplicasPresent
+    if (crossColoEnabled
         && diskDownCount + totalNotFoundCount > totalReplicaCount - diskReplicaSuccessTarget) {
-      logger.info(
-          "Terminating {} on {} due to disk down count and total Not_Found count. DiskDownCount: {}, TotalNotFoundCount: {}, TotalReplicaCount: {}, DiskReplicaSuccessTarget: {}",
-          routerOperation, partitionId, diskDownCount, totalNotFoundCount, totalReplicaCount, diskReplicaSuccessTarget);
-      routerMetrics.failedOnTotalNotFoundCount.inc();
-      return true;
+      if (!cloudReplicasPresent) {
+        logger.info(
+            "Terminating {} on {} due to disk down count and total Not_Found count. DiskDownCount: {}, TotalNotFoundCount: {}, TotalReplicaCount: {}, DiskReplicaSuccessTarget: {}",
+            routerOperation, partitionId, diskDownCount, totalNotFoundCount, totalReplicaCount, diskReplicaSuccessTarget);
+        routerMetrics.failedOnTotalNotFoundCount.inc();
+        return true;
+      } else {
+        // when all disk replicas are down, cloud returns Blob_Not_Found, we should return Blob_Not_Found, right?
+        // Jing TODO: What's the original reason to set it to false?
+        return true;
```

Comment:
Will wait for Casey's reply.

Reply:
Since a partition's replicas can span both disk replicas and cloud replicas for operations with cross colo retries, I don't know if we can easily subclass this and save code. (unless we structured our operationtracker a little differently and treated each datacenter as an individual unit and had a composite operationtracker for multicolo scenarios)

Reply:
Sounds good.